diffusion/models/text_encoder.py

```diff
@@ -0,0 +1,159 @@
+        encode_latents_in_fp16 (bool): Whether to encode text embeddings in fp16. Default: ``True``.
+        pretrained_sdxl (bool): Whether this text encoder is for a pretrained SDXL. If true, this will only use
+            the projected output from a CLIPTextModelWithProjection. Default: ``False``.
+    """
+
+    def __init__(
+        self,
+        model_names: Union[str, Tuple[str, ...]],
+        model_dim_keys: Optional[Union[str, List[str]]] = None,
+        encode_latents_in_fp16: bool = True,
+        pretrained_sdxl: bool = False,
+    ):
+        super().__init__()
+        self.pretrained_sdxl = pretrained_sdxl
+
+        if isinstance(model_names, str):
+            model_names = (model_names,)
+        if model_dim_keys is None:
+            model_dim_keys = ['projection_dim', 'd_model', 'hidden_size']  # CLIP, T5, E5
+        torch_dtype = torch.float16 if encode_latents_in_fp16 else None
+
+        self.text_encoders = torch.nn.ModuleList()
+        self.text_encoder_dim = 0
+        self.architectures = []
+        for model_name in model_names:
+            # Process model_name string and get model config
+            name_split = model_name.split('/')
+            base_name = '/'.join(name_split[:2])
+            subfolder = '/'.join(name_split[2:])
+            text_encoder_config = PretrainedConfig.get_config_dict(base_name, subfolder=subfolder)[0]
+
+            # Add text_encoder output dim to total dim
+            dim_found = False
+            for model_dim_key in model_dim_keys:
+                if model_dim_key in text_encoder_config:
+                    self.text_encoder_dim += text_encoder_config[model_dim_key]
+                    dim_found = True
+            if not dim_found:
+                raise ValueError(
+                    textwrap.dedent(f"""\
+                                    Did not find any model_dim_keys ({model_dim_keys}) in the config for {model_name}.\
+                                    Please specify the appropriate model_dim_keys for the target model config."""))
+
+            architectures = text_encoder_config['architectures']
+            if architectures == ['CLIPTextModel']:
+                self.text_encoders.append(
+                    CLIPTextModel.from_pretrained(base_name, subfolder=subfolder, torch_dtype=torch_dtype))
+            elif architectures == ['CLIPTextModelWithProjection']:
+                self.text_encoders.append(
+                    CLIPTextModelWithProjection.from_pretrained(base_name, subfolder=subfolder,
+                                                                torch_dtype=torch_dtype))
+            else:
+                self.text_encoders.append(
+                    AutoModel.from_pretrained(base_name, subfolder=subfolder, torch_dtype=torch_dtype))
+            self.architectures += architectures
+
+    @property
+    def device(self):
+        return self.text_encoders[0].device
+
+    def forward(self, tokenized_texts: torch.Tensor):
```

Comment:
Done! I had to add quiet a bit of masking code, but hopefully it is right now. It would be helpful if you can check my commits since your comment which will contain the masking code